go/vt/topo/tablet.go

```diff
@@ -503,41 +513,52 @@ func DeleteTabletReplicationData(ctx context.Context, ts *Server, tablet *topoda
 }
 
 // GetTabletMap tries to read all the tablets in the provided list,
-// and returns them all in a map.
-// If error is ErrPartialResult, the results in the dictionary are
+// and returns them in a map.
+// If error is ErrPartialResult, the results in the map are
 // incomplete, meaning some tablets couldn't be read.
 // The map is indexed by topoproto.TabletAliasString(tablet alias).
-func (ts *Server) GetTabletMap(ctx context.Context, tabletAliases []*topodatapb.TabletAlias) (map[string]*TabletInfo, error) {
+func (ts *Server) GetTabletMap(ctx context.Context, tabletAliases []*topodatapb.TabletAlias, opt *GetTabletsByCellOptions) (map[string]*TabletInfo, error) {
 	span, ctx := trace.NewSpan(ctx, "topo.GetTabletMap")
 	span.Annotate("num_tablets", len(tabletAliases))
 	defer span.Finish()
 
-	wg := sync.WaitGroup{}
-	mutex := sync.Mutex{}
+	var (
+		mu        sync.Mutex
+		wg        sync.WaitGroup
+		tabletMap = make(map[string]*TabletInfo)
+		returnErr error
+		// Previously this was always run with unlimited concurrency, so 32 should be fine.
+		concurrency = 32
+	)
 
-	tabletMap := make(map[string]*TabletInfo)
-	var someError error
+	if opt != nil && opt.Concurrency > 0 {
+		concurrency = opt.Concurrency
+	}
+	sem := make(chan int, concurrency)
 
 	for _, tabletAlias := range tabletAliases {
 		wg.Add(1)
 		go func(tabletAlias *topodatapb.TabletAlias) {
 			defer wg.Done()
+			sem <- 1
 			tabletInfo, err := ts.GetTablet(ctx, tabletAlias)
-			mutex.Lock()
+			<-sem
+			mu.Lock()
```

Comment:
I tried replacing this with FirstErrorRecorder. The code is no less awkward, because we don't actually want to record the returned error, we want to wrap it in a different error. And FirstErrorRecorder wraps its own mutex so you are not actually saving on that.

```
	var errRecorder concurrency.FirstErrorRecorder

	for _, tabletAlias := range tabletAliases {
		wg.Add(1)
		go func(tabletAlias *topodatapb.TabletAlias) {
			defer wg.Done()
			if err := sem.Acquire(ctx, 1); err != nil {
				// Only happens if context is cancelled.
				log.Warningf("%v: %v", tabletAlias, err)
				// We only need to set this on the first error.
				errRecorder.RecordError(NewError(PartialResult, tabletAlias.GetCell()))
				return
			}
			tabletInfo, err := ts.GetTablet(ctx, tabletAlias)
			sem.Release(1)
			if err != nil {
				log.Warningf("%v: %v", tabletAlias, err)
				// There can be data races removing nodes - ignore them for now.
				// We only need to set this on first error.
				if !IsErrType(err, NoNode) {
					errRecorder.RecordError(NewError(PartialResult, tabletAlias.GetCell()))
				}
			} else {
				mu.Lock()
				tabletMap[topoproto.TabletAliasString(tabletAlias)] = tabletInfo
				mu.Unlock()
			}
		}(tabletAlias)
	}
```

`sync.Once` is also problematic. The two places where we want to catch the error are slightly different and don't fit into a neat function. I had the same problem with trying to move the lock + logic + unlock into a closure.

Reply:
At least I was able to change both calls to Unlock to be deferred calls, so that's a win.